entity-framework/core/testing/index.md

```diff
@@ -1,117 +1,116 @@
 
-None of this is very important when testing EF Core internals because we use it specifically where the database is irrelevant to the test.
-On the other hand, these things tend to be very important when testing an application that uses EF Core.
+### Repository pattern
 
-## Unit testing
+The approaches above attempted to either swap EF Core's production database provider with a fake testing provider, or to stub out `DbSet` itself with an in-memory collection. These techniques are similar in that they evaluate the program's LINQ queries - either in SQLite or in memory - and this is ultimately the source of the difficulties outlined above: a query designed to run against a specific production database cannot run elsewhere without issues.
+
+For proper, reliable unit testing, consider introducing a [repository layer](https://martinfowler.com/eaaCatalog/repository.html) which mediates between your application code and EF Core. The production implementation of the repository contains the actual LINQ queries and executes them via EF Core. In testing, the repository abstraction is directly stubbed or mocked without being needing any actual LINQ queries, effectively removing EF Core from your testing stack altogether and allowing tests to focus on application code alone.
+
+The following diagram compares the database fake pattern (SQLite/InMemory) with the repository pattern:
+
+![Comparison of fake provider with repository pattern](_static/fake-provider-and-repository-pattern.png)
+
+Since LINQ queries are no longer part of testing, you can directly provide query results to your application. Put another way, the previous approaches roughly allow stubbing out *query inputs* (e.g. replacing SQL Server tables with an in-memory one), but then still execute the actual query operators in-memory; the repository pattern, in contrast, allows you to stub out *query outputs* directly, allowing for far more powerful and focused unit testing.
+
+One possible disadvantage of the repository pattern is that it requires you to architect your application accordingly: application code can no longer execute LINQ queries directly, but must instead call into a repository method, which executes the query (and is stubbed in testing). This can make for a somewhat heavier application architecture, but can also have other advantages: all data access code is concentrated in one place rather than being spread around the application, and if your application needs to support more than one database, the repository abstraction can be very helpful for tweaking queries across providers.
```

Comment:
I think it's important to point out that not making IQueryable available to your application has _very significant_ disadvantages. IQueryable, and that way it can be used to compose complex queries, is, I think, one of the things that makes LINQ and C#/.NET great. If you hide all that behind a repository, then you have introduce your own abstractions for creating complex queries. This is cleaner in terms of separation of concerns and not letting your persistence layer leak into your domain layer, but is also consequentially _much more limiting_ (because by-design a repository limits the queries that can be constructed) and _much more work_. If you're doing this only for the dubious benefit of unit-testing your persistence layer, then I think very often you will be making the wrong tradeoff.

It's also very important to emphasize that repository is only beneficial if it really does abstract way queries into well-defined, domain specific requests of the repository. This is why the "generic repository" pattern is generally pointless--especially if exposes IQueryables directly, but even if it just replaces IQueryable with another way to build general-purpose queries that may behave differently on different providers.

Reply:
I may want the repo to handle business logic for C.U.D. operations but still have the flexibility to build a query based on criteria. The repo can still manage the connection, context, etc. but if it exposes queries like this, you can still test without a live database and build logic into your repo to approve the expression tree before it's used.

```csharp
// signature 
Task<ICollection<T>> Query<T>(Func<IQueryable<T>, IQueryable<T>> filter);

// repo
using var ctx = new Context();
return await filter(ctx.Set<T>()).ToListAsync();

// consumer
var sold = await repo.Query<Item>(q => q.Where(i => i.IsSold));
```

Reply:
That's an interesting point. We're now discussing certain "dynamic" scenarios where e.g. the web API itself exposes flexible querying, right?

First, my unfounded intuition is that flexible, dynamic query APIs are on the more rare/advanced side. I think a lot of scenarios have fixed, preset query shapes which can easily be abstracted behind repository methods with regular parameters. Limited dynamic scenarios can be implemented without full-blown IQueryable by having the repository can expose a simplified, non-queryable API (e.g. a string parameter with the name of the property to order by).

Now, for cases where a full IQueryable really is needed (totally arbitrary filterable/sortable/projectable web API, again - seems rare)... Users could expose IQueryable here, and have their repository mock perform verification that the given query tree has the expected shape. I don't think this is fundamentally different from mock-style verification of other parameter types - just more elaborate/complex. So I don't think the interdiction against exposing IQueryable from repositories has to be absolute; it's just good practice to avoid it unless really needed.

> It's also very important to emphasize that repository is only beneficial if it really does abstract way queries into well-defined, domain specific requests of the repository [...]

I completely agree. I was mainly trying to avoid going into the *general* repository discussion here, and to only talk about repositories as they relate to testing (this paragraph goes a bit beyond that).